With a large dataset of tweets and a wide range of possible cities, we wanted to confirm that cities on people's accounts matched a list of all cities in the United States. 

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', 1000)

In [2]:
df = pd.read_csv('./combined_us.csv')
cities = pd.read_csv('./USA-cities-and-states/us_cities_states_counties.csv', sep = '|')

In [3]:
df

status_id             user_id                 created_at  \
0       1255648063143710723            92677101  2020-04-30 00:00:01+00:00   
1       1255648072090169345           217987384  2020-04-30 00:00:03+00:00   
2       1255648079790903297            62657173  2020-04-30 00:00:05+00:00   
3       1255648164809445382  988863378427990017  2020-04-30 00:00:25+00:00   
4       1255648197835251713           475393715  2020-04-30 00:00:33+00:00   
...                     ...                 ...                        ...   
155374  1244413925912907778           160304538  2020-03-29 23:59:34+00:00   
155375  1244413952945250305  862769187499986944  2020-03-29 23:59:41+00:00   
155376  1244413955172372481            80139649  2020-03-29 23:59:41+00:00   
155377  1244414009736118277            32232449  2020-03-29 23:59:54+00:00   
155378  1244414009329082368           113628427  2020-03-29 23:59:54+00:00   

            screen_name                                               text  \
0           lopezgovlaw  #SurveySunday: Most Miamians Think #Coronaviru...   
1         maisondecorla  *Breaking News* This just in - Stonewall Kitch...   
2           drosssports  Once again, while it wasn’t @DrDavidKatz , @to...   
3            prana_mani  Free #webinar announcement! This Sunday, May 3...   
4            KPobgyndoc  Went on a #COVID19 walk with my #mentor on her...   
...                 ...                                                ...   
155374        iskandrah  I don't know who needs to hear this, but #coro...   
155375  Customer_Effort  Trying my #comfilife #coccyx #cushion #sciatic...   
155376  rickeybevington  I am surprised too &gt; “Local leaders and som...   
155377   djfenixchicago  Another 30! #socialdistancing #coronavirus #co...   
155378         isamh195  #Covid_19 relief. Better things to talk about ...   

                     source  is_quote  is_retweet  favourites_count  \
0        Twitter for iPhone     False       False              8132   
1                 Instagram     False       False                67   
2        Twitter for iPhone      True       False             38651   
3        Twitter for iPhone     False       False               159   
4        Twitter for iPhone     False       False              6336   
...                     ...       ...         ...               ...   
155374  Twitter for Android     False       False            171376   
155375   Twitter for iPhone     False       False                95   
155376     Twitter for iPad      True       False             15021   
155377  Twitter for Android     False       False              1077   
155378  Twitter for Android     False       False               244   

        retweet_count  ...  place_full_name place_type followers_count  \
0                   2  ...        Miami, FL       city           16007   
1                   0  ...       Boutte, LA       city             105   
2                   0  ...      Chicago, IL       city            6441   
3                   1  ...         Troy, NY       city              23   
4                   2  ...      Oakland, CA       city            7161   
...               ...  ...              ...        ...             ...   
155374              4  ...    Illinois, USA      admin           18163   
155375              0  ...   Minnesota, USA      admin              29   
155376              1  ...      Atlanta, GA       city            4199   
155377              0  ...      Chicago, IL       city             819   
155378              0  ...  Albuquerque, NM       city             131   

        friends_count         account_created_at verified  lang is_reply  \
0                3634  2009-11-26 03:42:10+00:00     True    en    False   
1                 146  2010-11-21 03:03:54+00:00    False    en    False   
2                 501  2009-08-03 23:44:39+00:00     True    en    False   
3                  93  2018-04-24 19:32:56+00:00    False    en    False   
4                6670 

In [4]:
cities

City State short   State full                 County  \
0       Holtsville          NY     New York                SUFFOLK   
1       Holtsville          NY     New York                SUFFOLK   
2         Adjuntas          PR  Puerto Rico               ADJUNTAS   
3         Adjuntas          PR  Puerto Rico               ADJUNTAS   
4         Adjuntas          PR  Puerto Rico               ADJUNTAS   
...            ...         ...          ...                    ...   
63206      Klawock          AK       Alaska  PRINCE OF WALES HYDER   
63207   Metlakatla          AK       Alaska  PRINCE OF WALES HYDER   
63208  Point Baker          AK       Alaska  PRINCE OF WALES HYDER   
63209    Ward Cove          AK       Alaska      KETCHIKAN GATEWAY   
63210     Wrangell          AK       Alaska               WRANGELL   

                     City alias  
0      Internal Revenue Service  
1                    Holtsville  
2               URB San Joaquin  
3              Jard De Adjuntas  
4           Colinas Del Gigante  
...                         ...  
63206                   Klawock  
63207                Metlakatla  
63208               Point Baker  
63209                 Ward Cove  
63210                  Wrangell  

[63211 rows x 5 columns]

The cities dataset contains a list of every city, state, county, and many alternative city names in the United States. The dataset was put together by grammakov and can be found unaltered on their [github](https://github.com/grammakov/USA-cities-and-states)

In [5]:
city = list(cities['City'])
city = list(set(city))

In [6]:
del city[0]

In [9]:
df_c = list(df['city'])
df_c = list(set(df_c))

Each list or column of city names was converted to a set in order to compare overlapping city names without taking up an immense amount of time iterating over hundred or thousands of duplicates. 

In [10]:
len(df_c)

5628

There were 5628 unique city names used in our twitter dataset. 

In [11]:
# https://medium.com/better-programming/a-visual-guide-to-set-comparisons-in-python-6ab7edb9ec41
inter = set(df_c).intersection(set(city))

In [12]:
inter

{'Delray Beach',
 'Bath',
 'Lemoyne',
 'Redford',
 'Jupiter',
 'Chalmette',
 'Port Lavaca',
 'Altadena',
 'Mukilteo',
 'Mendota',
 'Litchfield Park',
 'Southbridge',
 'Thatcher',
 'Lake Elsinore',
 'Oxford',
 'Wellington',
 'Biddeford',
 'Shelbyville',
 'Port Chester',
 'Pepperell',
 'Lamar',
 'Windham',
 'Roanoke',
 'Herculaneum',
 'Wellesley',
 'Hemet',
 'Marietta',
 'Elizabeth',
 'Hagerstown',
 'Dryden',
 'South Elgin',
 'Sunland Park',
 'Westport',
 'Mountain Home',
 'Sunset Beach',
 'Redan',
 'Groveport',
 'Hurricane',
 'Ipswich',
 'Muldraugh',
 'North Palm Beach',
 'Selinsgrove',
 'Port Salerno',
 'South Sioux City',
 'Goshen',
 'Lynwood',
 'Harrisonburg',
 'Bayville',
 'Justice',
 'Syracuse',
 'Suffolk',
 'Fernandina Beach',
 'Chino',
 'Three Oaks',
 'Santa Cruz',
 'Addison',
 'Severn',
 'Pelham',
 'Pine Island',
 'Reisterstown',
 'Hanover',
 'Weslaco',
 'Crown Point',
 'Picayune',
 'Howard',
 'Pflugerville',
 'Swansea',
 'Wyncote',
 'Brackenridge',
 'Granger',
 'Galt',
 'San Le

In [13]:
len(inter)

4159

Inter shows all cities that are present in both the city locations listed on Twitter and cities in the master list of US cities. With only 4159 cities overlapping, there is clearly some discrepency between the lists on twitter and the master list of cities. 

In [14]:
no_intersection = list(set(df_c) - set(city))

In [15]:
len(no_intersection)

1469

In [16]:
no_intersection

['Progress',
 'Grosse Pointe Farms',
 'Surfside Beach',
 'Orland Hills',
 'Fairwood',
 'King of Prussia',
 'Middleborough Center',
 'Blasdell',
 'Westwood Shores',
 'Hide-A-Way Lake',
 'Belville',
 'Dix Hills',
 'Indian Shores',
 'Plandome',
 'North St Paul',
 'Scott Township',
 'Ingalls Park',
 'Shoreline',
 'Country Homes',
 'McKee City',
 'Buckhall',
 'Delhi Hills',
 'Madeira',
 'North Shore',
 'Grandview Heights',
 'Dargin',
 'Big Coppitt Key',
 'Fort Carson',
 'Orange Blossom Hills',
 'Lake Success',
 'Security',
 'McLean',
 'Blennerhassett',
 'Harahan',
 'Clyde Hill',
 'Green Park',
 'Neptune City',
 'Wagon Wheel',
 'Farmington Hills',
 'Bull Run',
 'Walker Mill',
 'Howland Center',
 'Avilla Beach',
 'Agua Fria',
 'Bells Cross Roads',
 'Blue Ash',
 'Ocean Pines',
 'Sunny Isles Beach',
 'Lutherville',
 'Oroville East',
 'Hoover',
 'Dale City',
 'Kendall West',
 'Dakota Ridge',
 'St Matthews',
 'Branchburg',
 'Pasatiempo',
 'Phalanx',
 'Lynnwood-Pricedale',
 'Picnic Point',
 'North

No_intersection shows every city listed as a location on twitter that was not found in our master list of US cities. Looking through a number of these 'cities' that were found on Twitter but not on the master list revealed that many were either neighborhoods or 'census designated places'. 'Census designated places' or cdp's are unincorporated communities who may have a colloquial name but only legally exist for census collection purposes. They are not self-governing municipalities and thus would not appear on a list of US cities. Likewise, neighborhoods of major cities may act as common reference for where someone lives, but they rarely have their own administrations separate from the cities they are located within. 

In [17]:
check = set(city).intersection(no_intersection)

In [18]:
check

set()

We no longer have any intersecting cities between our two sets of cities. 